src/app/shared/dovetail/dovetail-audio.ts

```diff
@@ -333,52 +312,40 @@ export class DovetailAudio extends ExtendableAudio {
   }
 
   private skipToFile(index: number, resume: false|PromiseResolver<void> = false): boolean {
-    if (this.index !== index && this.arrangement.entries.length > index) {
+    if (this.index !== index && this.arrangement.length > index) {
       if (this.index !== -1) {
-        const eventData: SegmentEventData = {
-          segment: this.arrangement.entries[this.index],
-          segmentType: this.arrangement.entries[this.index].type
-        };
-        this.$$sendEvent(SEGMENT_END, eventData);
+        this.$$sendEvent(SEGMENT_END, {segment: this.arrangement[this.index]});
       }
       this.index = index;
-      let arrangement = this.arrangement.entries[index];
-      this.$$debug(`Goto: ${arrangement.id}`);
+      let nextSegment = this.arrangement[index];
+      this.$$debug(`Goto: ${nextSegment.id}`);
 
       // skip unplayable / 0-length audio
-      if (arrangement.unplayable) {
+      if (nextSegment.unplayable) {
         return this.skipToFile(index + 1, resume);
       }
 
-      this._audio.src = arrangement.audioUrl;
+      this._audio.src = nextSegment.audioUrl;
       this._audio.playbackRate = this.playbackRate;
       if (resume) { resume(this.playItSafe()); }
 
-      this.$$sendEvent(SEGMENT_START, {
-        segment: arrangement,
-        segmentType: arrangement.type
-      });
+      this.$$sendEvent(SEGMENT_START, {segment: nextSegment});
 
       return true;
     } else {
       return false;
     }
   }
 
-  private $$logImpression(event: SegmentEventData) {
-    const type: string = event.segmentType;
-    const segment: DovetailArrangementEntry = event.segment;
-
-    if (type === 'ad' || type === 'houseAd') {
-      if (typeof this._imgElem === 'undefined') {
-          this._imgElem = document.createElement('img');
-          this._imgElem.width = 1;
-          this._imgElem.height = 1;
-          this._imgElem.style.position = 'absolute';
-          document.body.appendChild(this._imgElem);
-      }
-      this.$$debug(`Impress: ${segment.id}`);
-      this._imgElem.src = segment.impressionUrl;
+  private $$logDownload(event: DovetailSegmentEvent) {
+    if (event.segment.trackBefore()) {
```

Comment:
These track functions return the number of pixel-trackers hit.  So if it's > 0, the next line will debug log to the console.  (Assuming you have `?debug` in your url somewhere).

Reply:
so currently beforeTrackers is just for dovetail's pixel tracker, which we'll put in there once as part of the construction of the first segment? and then the afterTrackers will be one for each ad?

Reply:
Correct.  And eventually, we'll add 3rd-party pingbacks (which come out of the Adzerk metadata) into the `afterTrackers` as well.